phylovi/libsbn
# 234 4 uniform toplogy final

## Description

+ Changes all likelihood calculations to be product-sum as we found out that GP does not compute sum-product. 
+ `GPEngine` maintains log likelihood as matrix, with the number of rows equal to number of GPCSPs and the number of columns equal to the number of sites.
+ All subsequent computation, particularly `GPEngine::UpdateSBNProbabilities` and `GPInstance::EstimateSBNParameters` code that uses the log likelihood computation has been updated.

Fixes issue #235 -- cherry picked commits to 235-optimization by @hrnasif (see 7140dec018c056c74802fa7158b7c28bcb96f479).

## Tests
All tests are in `src/gp_doctest.cpp`.
+ Wrote function for computing exact prod-sum likelihood given unrooted newick and fasta.
+ To implement above, we allowed extracting a single site from `Alignment` instance and a setter to supply `Alignment` object to `GenericSBNInstance` instance.
+ Fixed and strengthened old tests to compare against exact prod-sum likelihood.
+ Added marginal likelihood test on five taxa data involving 4 trees.
+ Ensure PLVs are populated correctly during `EstimateBranchLengths`.
+ Ensure SBN parameters are updated correctly.

## Checklist:

* [x] `make format` has been run
* [x] The code uses informative and accurate variable and function names
* [x] The functionality is factored out into functions and methods with logical interfaces
* [x] Comments are up to date, document intent, and there are no commented-out code blocks
* [x] Commenting and/or documentation is sufficient for others to be able to understand intent and implementation
* [x] Variables are named according to our conventions
* [x] `const` used where appropriate
* [x] The code uses modern C++ conventions, including range-for, `auto`, and structured bindings
* [x] TODOs have been eliminated from the code
